### PR TITLE
[TRTLLM-11421][fix] fix data races in speculative decoding fast-logits handoff

### PR DESCRIPTION
**Pull request by eopXD**
## Summary

- `draftModelSendLogitsThread` called `terminateRequest()` directly from a background thread, racing with the main thread on `KVCacheManager::removeSequence()` and other shared batch manager state (sequence slot manager, PEFT cache manager).
- Fix: after sending logits via MPI, the background thread now pushes completed requests into a handoff queue (`mDraftRequestsDoneSendingLogits`) drained by the main thread in `forwardSync()`, where `terminateRequest()` is called.
- The same two threads also shared `mDraftRequestsWaitingToSendLogits` (main thread push_back, background thread iterate+erase) without synchronization. Fix: extend the existing mutex (renamed `mDraftRequestsDoneMtx` → `mDraftRequestsMtx`) to guard both queues. MPI sends remain outside the lock to avoid serializing network I/O.
- This also removes the logits thread dependency on `SequenceSlotManager`, `BaseKVCacheManager`, and `BasePeftCacheManager`, simplifying its interface.

## Related

- Complements #12251 (KVCacheManager TOCTOU race fixes) -- that PR fixes races within the KV cache manager itself; this PR fixes the upstream race where `terminateRequest` was called from the wrong thread entirely.

## Test plan

- [x] Existing speculative decoding fast-logits orchestration tests pass
- [x] Stress test with overlap scheduling / MTP mode to exercise the concurrent send-logits + terminate path
- [x] CI: `/bot run`

## Summary by CodeRabbit

* **Refactor**
  * Refactored speculative decoding thread coordination: request termination is now deferred from the background logits thread to the main thread via a shared synchronization queue, improving inter-thread synchronization and simplifying background thread dependencies.

### CODE REVIEW

**Comment by SimengLiu-nv** [comment]
Codex review points out a potential race condition:

The race is between two threads mutating the same std::vector without a happens-before relationship.

In the main batch-manager thread, updateRequests() appends completed draft requests here:

``` c++
// trtGptModelInflightBatching.cpp (line 2527)

mDraftRequestsWaitingToSendLogits.push_back(llmReq);
```

At the same time, the background MPI sender thread scans and erases from that same vector here:

``` c++
// logitsThread.cpp (line 90)

for (auto it = draftRequestsWaitingToSendLogits->begin();
     it != draftRequestsWaitingToSendLogits->end(); ++it)
{
    ...
    draftRequestsWaitingToSendLogits->erase(it);
}
```
A concrete interleaving:

Draft request A finishes on the draft model. The main thread enters updateRequests() and is about to append A to mDraftRequestsWaitingToSendLogits.

The target model concurrently calls targetModelReceiveLogits() and sends an MPI request asking the draft model for logits for an earlier request B.

draftModelSendLogitsThread() receives that MPI request and starts iterating mDraftRequestsWaitingToSendLogits to find B.

While the sender thread holds an iterator into the vector, the main thread executes push_back(A).

If push_back() reallocates, the vector moves its storage and invalidates the sender thread’s iterator. The sender can then dereference freed/moved storage or erase using an invalid iterator.

Even if no reallocation happens, push_back() still mutates vector size and constructs a new element while the sender thread is reading begin(), end(), and elements. That is still a C++ data race and therefore undefined behavior.

There is also the reverse mutation case: the sender thread may erase() an element while the main thread is appending. erase() shifts elements and updates vector size; push_back() may update size/capacity at the same time. That can corrupt vector metadata, drop a request, duplicate a request pointer, or make the later TLLM_CHECK(draftRequest != nullptr) fail spuriously.

The PR fixes the later “done sending, now terminate” handoff with a mutex, but the earlier “waiting to send logits” handoff remains unsynchronized.

**Review by SimengLiu-nv — CHANGES_REQUESTED** [review]
Need to verify potential race conditions.

**Comment by eopXD** [comment]
Addressed with latest commit. Now the four access sites, all now under mDraftRequestsMtx:

- trtGptModelInflightBatching.cpp:2540 — push_back to waiting queue (main thread, new lock)
- logitsThread.cpp:100 — iterate+erase on waiting queue (sender thread, new lock)
- logitsThread.cpp:119 — push_back to done queue (sender thread, existing lock, mutex renamed)
- trtGptModelInflightBatching.cpp:911 — swap of done queue (main thread, existing lock, mutex renamed)

**Review by SimengLiu-nv — APPROVED** [review]
LGTM.

**Comment by eopXD** [comment]
/bot --skip "We had earned success before. x86 Single GPU has passed. Failure in the latest CI/CD is not related to our code change. Let us merge the MR without further delay."

**Comment by eopXD** [comment]
/bot skip --comment "We had earned CI/CD success before. In the latest CI/CD, x86 Single GPU has passed. The failure is not related to our code change. Let us merge the MR without further delay."
